src/geometry.ts

```diff
@@ -0,0 +1,304 @@
+    height?: number;
+    left?: number;
+    top?: number;
+    width?: number;
+}
+
+/**
+ * Returns an object with information on the dimensions of an Element's border box.
+ * @param element The Element to measure
+ * @return An object reporting the top, right, bottom, left, width, and height of the Element's border box
+ */
+export function getBorderBox(element: Element): ClientRect {
+	// Return a plain/mutable object (not a native read-only ClientRect) to be consistent with the other functions.
+	// We may end up modifying these values anyway (for scroll adjustment).
+	return adjustBox(element.getBoundingClientRect(), 0, 0, 0, 0);
+}
+
+/**
+ * Returns an object with information on the dimensions of an Element's content box.
+ * @param element The Element to measure
+ * @return An object reporting the top, right, bottom, left, width, and height of the Element's content box
+ */
+export function getContentBox(element: Element): ClientRect {
+	// While this could reuse getPaddingBox, it does not, to avoid an extra getComputedStyle call.
+	const style = getComputedStyle(element);
+	return adjustBox(getBorderBox(element),
+		-pxToNumber(style.borderTopWidth) - pxToNumber(style.paddingTop),
+		-pxToNumber(style.borderRightWidth) - pxToNumber(style.paddingRight),
+		-pxToNumber(style.borderBottomWidth) - pxToNumber(style.paddingBottom),
+		-pxToNumber(style.borderLeftWidth) - pxToNumber(style.paddingLeft)
+	);
+}
+
+/**
+ * Returns an object with information on the dimensions of an Element's margin box.
+ * @param element The Element to measure
+ * @return An object reporting the top, right, bottom, left, width, and height of the Element's margin box
+ */
+export function getMarginBox(element: Element): ClientRect {
+	const style = getComputedStyle(element);
+	return adjustBox(getBorderBox(element),
+		pxToNumber(style.marginTop),
+		pxToNumber(style.marginRight),
+		pxToNumber(style.marginBottom),
+		pxToNumber(style.marginLeft)
+	);
+}
+
+/**
+ * Returns an object with information on the dimensions of an Element's padding box.
+ * @param element The Element to measure
+ * @return An object reporting the top, right, bottom, left, width, and height of the Element's padding box
+ */
+export function getPaddingBox(element: Element): ClientRect {
+	const style = getComputedStyle(element);
+	return adjustBox(getBorderBox(element),
+		-pxToNumber(style.borderTopWidth),
+		-pxToNumber(style.borderRightWidth),
+		-pxToNumber(style.borderBottomWidth),
+		-pxToNumber(style.borderLeftWidth)
+	);
+}
+
+/**
+ * Sets the dimensions of an Element's border box.
+ * @param element The Element to modify
+ * @param rect An object containing optional number properties for height, left, top, and width
+ */
+export function setBorderBox(element: HTMLElement, rect: SettableClientRect): void {
+	const style = getComputedStyle(element);
+	const borderBox = style.boxSizing && style.boxSizing === 'border-box';
+
+	if (rect.height) {
```

Comment:
This will be false if `height` is 0; we should check `if ('height' in rect)` instead.

(This is also applicable to width/top/left and in all 4 of these functions.  If I'm reading dojo 1's code correctly, it wouldn't have handled this right either for width/height)